### PR TITLE
Add basic MMU (and supervisor mode) support for RISC-V

### PR DESCRIPTION
**Pull request by berendo**
This PR implements the foundational changes for running the Zephyr kernel in supervisor mode and enabling the MMU. However, full user space thread support is not yet implemented, as the integration between the MMU implementation and the Zephyr memory protection and region support is still work in progress. Also, the only supported page size is 2 MB (except for the lowest 4 KB page for catching NULL pointer dereferences).

This PR also includes a couple of cherry-picked commits that are required RISC-V bug fixes for MPU (MMU) protection from the Zephyr upstream `main` that our `recogni` branch currently lacks. These will melt away when we rebase our `recogni` branch at a future time.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -624,7 +624,7 @@ if(CONFIG_TIMEOUT_64BIT)
   set(SYSCALL_SPLIT_TIMEOUT_ARG --split-type k_timeout_t)
 endif()
 
-if(CONFIG_BOARD_SCORPIO)
+if(CONFIG_BOARD_SCORPIO AND NOT CONFIG_RISCV_MMU)
```
Comment (sabhiram) [inline CMakeLists.txt:627]: I didnt know you that you can do `AND NOT` :)

Comment (berendo) [reply CMakeLists.txt:627]: Yup! 😄

File: arch/riscv/core/isr.S [hunk 2]
```diff
@@ -202,6 +202,20 @@
 	DO_CALLER_SAVED(RV_OP_LOADREG)		;\
 	addi sp, sp, __z_arch_esf_t_SIZEOF	;
 
+#if defined(CONFIG_RISCV_SUPERVISOR)
```
Comment (sabhiram) [inline arch/riscv/core/isr.S:205]: Nice, that is so easy to follow :)

**Review by sabhiram — APPROVED** [review]
Looks good on first glance. Need to understand isr_machine and the associated things which ill pull the code and follow with more than just diffs. 

Amazing that this is actually here! 🥇

File: arch/riscv/core/isr.S (hunk 2 above)
Comment (berendo) [reply arch/riscv/core/isr.S:205]: 👍

**Comment by berendo** [comment]
> Need to understand isr_machine and the associated things which ill pull the code and follow with more than just diffs.

Yes, I intend to do an overview, but first I have to get some materials together to present an overview of the RISC-V privileged architecture, especially vis a vis machine mode to supervisor mode delegation of exceptions and interrupts. At the same time, I'd like to get the baseline Zephyr changes (which have no effect until `CONFIG_RISV_MMU` is enabled) merged, so that I can avoid an endless loop of cleaning up...

**Comment by brudley** [comment]
Nice. Looking forward to your overview!

**Pull request merged**